File: backend/ms-users/src/main/java/pi/ms_users/domain/Contract.java
package pi.ms_users.domain;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@AllArgsConstructor
@NoArgsConstructor
@Data
@Entity
@Table(name = "Contract")
public class Contract {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "user_id", nullable = false)
    private String userId;

    @Column(name = "property_id", nullable = true)
    private Long propertyId;

    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false)
    private ContractType contractType;

    @Column(name = "start_date", nullable = false)
    private LocalDate startDate;

    @Column(name = "end_date", nullable = false)
    private LocalDate endDate;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private ContractStatus contractStatus;

    @Enumerated(EnumType.STRING)
    @Column(name = "currency", nullable = false)
    private PaymentCurrency currency;

    @Column(name = "initial_amount", nullable = false)
    private BigDecimal initialAmount;

    @Column(name = "adjustment_frequency_months", nullable = false)
    private Integer adjustmentFrequencyMonths;

    @Column(name = "last_paid_amount", nullable = true)
    private BigDecimal lastPaidAmount;

    @Column(name = "last_paid_date", nullable = true)
    private LocalDateTime lastPaidDate;

    @Column(name = "note", nullable = true)
    private String note;

    @Column(name = "has_deposit", nullable = false)
    private boolean hasDeposit = false;

    @Column(name = "deposit_amount", precision = 15, scale = 2)
    private BigDecimal depositAmount;    

    @Column(name = "deposit_note", length = 1000)
    private String depositNote;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "adjustment_index_id", nullable = false, foreignKey = @ForeignKey(name = "fk_contract_adjustment_index"))
    private IncreaseIndex adjustmentIndex;

    @OneToMany(mappedBy = "contract", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<ContractUtility> contractUtilities = new HashSet<>();

    @OneToMany(mappedBy = "contract", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<ContractIncrease> contractIncrease = new HashSet<>();

    @OneToOne(mappedBy = "contract", cascade = CascadeType.ALL, orphanRemoval = true)
    private Commission commission;

    @OneToMany(mappedBy = "contract", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<Payment> payments = new HashSet<>();

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(name = "Contract_Guarantor", joinColumns = @JoinColumn(name = "contract_id"), inverseJoinColumns = @JoinColumn(name = "guarantor_id"))
    private Set<Guarantor> guarantors = new HashSet<>();

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Contract c)) return false;
        return id != null && id.equals(c.id);
    }

    @Override
    public int hashCode() {
        return 31;
    }
}

File: frontend/src/app/user/types/contract.ts
import type { Payment, PaymentCurrency } from "./payment";
import type { IncreaseIndex } from "./increaseIndex";
import { ContractUtilityGet } from "./contractUtility";
import { ContractIncrease } from "./contractIncrease";
import { Commission } from "./commission";

export enum ContractType {
  TEMPORAL = "TEMPORAL",
  VIVIENDA = "VIVIENDA",
  COMERCIAL = "COMERCIAL",
}

export enum ContractStatus {
  ACTIVO = "ACTIVO",
  INACTIVO = "INACTIVO",
}

// DTO base para crear/actualizar
export interface Contract {
  id: number;
  userId: string;
  propertyId: number;
  contractType: ContractType;
  startDate: string; // ISO date
  endDate: string; // ISO date
  contractStatus: ContractStatus;
  currency: PaymentCurrency;
  initialAmount: number;
  adjustmentFrequencyMonths: number;
  // Compatibilidad hacia atrás con UI previa
  increase?: number;
  increaseFrequency?: number;
  lastPaidAmount: number | null;
  lastPaidDate: string | null; // ISO date-time
  note: string | null;
  hasDeposit: boolean;
  depositAmount: number | null;
  depositNote: string | null;
  adjustmentIndexId: number | null;
  guarantorsIds?: number[];
}

export type ContractCreate = Omit<Contract, "id">;

// Respuesta enriquecida (ContractGetDTO)
export interface ContractGet {
  id: number;
  userId: string;
  propertyId: number;
  contractType: ContractType;
  startDate: string;
  endDate: string;
  contractStatus: ContractStatus;
  currency: PaymentCurrency;
  initialAmount: number;
  adjustmentFrequencyMonths: number;
  lastPaidAmount: number | null;
  lastPaidDate: string | null;
  note: string | null;
  hasDeposit: boolean;
  depositAmount: number | null;
  depositNote: string | null;
  adjustmentIndex: IncreaseIndex;
  contractUtilities: ContractUtilityGet[];
  contractIncrease: ContractIncrease[];
  commission: Commission | null;
  payments: Payment[];
  guarantors: Array<{ id: number; name: string; phone: string; email: string }>;
}

// Representación mínima de contrato para listados simples
export interface ContractSimple {
  id: number;
  userId: string;
  propertyId: number;
  contractType: ContractType;
  startDate: string;
  endDate: string;
  contractStatus: ContractStatus;
}
